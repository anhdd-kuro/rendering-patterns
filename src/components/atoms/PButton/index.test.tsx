import { fireEvent, render, screen } from "@testing-library/react";
import { PButton } from ".";

describe("Atoms/PButton", () => {
  describe("Button disabled", () => {
    const mockedOnClick = vi.fn();

    let disabledButton: HTMLButtonElement, normalButton: HTMLButtonElement;

    beforeEach(() => {
      mockedOnClick.mockReset();

      render(
        <>
          <PButton disabled onClick={mockedOnClick}>
            Disabled Button
          </PButton>
          <PButton onClick={mockedOnClick}>Normal Button</PButton>
        </>,
      );

      disabledButton = screen.getByText("Disabled Button");
      normalButton = screen.getByText("Normal Button");
    });

    afterEach(() => {
      expect(disabledButton).matchSnapshot();
      expect(disabledButton).matchSnapshot();
    });

    test("render", () => {
      expect(disabledButton).toBeDefined();
      expect(disabledButton.tagName).toBe("BUTTON");
    });

    test("attributes", () => {
      expect(disabledButton.getAttribute("disabled")).toBe("");
      expect(normalButton.getAttribute("disabled")).toBeNull();
    });

    test("click event", () => {
      fireEvent.click(disabledButton);
      expect(mockedOnClick).not.toBeCalled();

      fireEvent.click(normalButton);
      expect(mockedOnClick).toBeCalledTimes(1);
    });
  });

  describe("Button type", () => {
    let button: HTMLButtonElement;

    afterEach(() => {
      expect(button).matchSnapshot();
    });

    test("should have button default type", () => {
      render(<PButton>Button</PButton>);
      button = screen.getByText("Button");
      expect(button.getAttribute("type")).toBe("button");

      expect(button).matchSnapshot();
    });

    test("should have given type", () => {
      render(<PButton type="submit">Typed Button</PButton>);
      button = screen.getByText("Typed Button");
      expect(button.getAttribute("type")).toBe("submit");
    });
  });
});
