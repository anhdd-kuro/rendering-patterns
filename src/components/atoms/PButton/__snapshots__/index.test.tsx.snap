// Vitest Snapshot v1

exports[`Atoms/PButton > Button disabled > attributes 1`] = `
<button
  class="flex-center opacity-60 cursor-not-allowed border border-gray-300 false"
  disabled=""
  type="button"
>
  Disabled Button
</button>
`;

exports[`Atoms/PButton > Button disabled > attributes 2`] = `
<button
  class="flex-center opacity-60 cursor-not-allowed border border-gray-300 false"
  disabled=""
  type="button"
>
  Disabled Button
</button>
`;

exports[`Atoms/PButton > Button disabled > click event 1`] = `
<button
  class="flex-center opacity-60 cursor-not-allowed border border-gray-300 false"
  disabled=""
  type="button"
>
  Disabled Button
</button>
`;

exports[`Atoms/PButton > Button disabled > click event 2`] = `
<button
  class="flex-center opacity-60 cursor-not-allowed border border-gray-300 false"
  disabled=""
  type="button"
>
  Disabled Button
</button>
`;

exports[`Atoms/PButton > Button disabled > render 1`] = `
<button
  class="flex-center opacity-60 cursor-not-allowed border border-gray-300 false"
  disabled=""
  type="button"
>
  Disabled Button
</button>
`;

exports[`Atoms/PButton > Button disabled > render 2`] = `
<button
  class="flex-center opacity-60 cursor-not-allowed border border-gray-300 false"
  disabled=""
  type="button"
>
  Disabled Button
</button>
`;

exports[`Atoms/PButton > Button type > should have button default type 1`] = `
<button
  class="flex-center false border border-gray-300 false"
  type="button"
>
  Button
</button>
`;

exports[`Atoms/PButton > Button type > should have button default type 2`] = `
<button
  class="flex-center false border border-gray-300 false"
  type="button"
>
  Button
</button>
`;

exports[`Atoms/PButton > Button type > should have given type 1`] = `
<button
  class="flex-center false border border-gray-300 false"
  type="submit"
>
  Typed Button
</button>
`;
